File: driver/php/ext/class/class_cursor.php
<?php
/*******************************************************************************
   Copyright (C) 2012-2014 SequoiaDB Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*******************************************************************************/

/** \file class_cursor.php
    \brief cursor class
 */

/**
 * SequoiaCursor Class. To get this Class object must be call SequoiaDB::list or SequoiaDB::snapshot or SequoiaCL::find, and so on.
 */
class SequoiaCursor
{
   /**
    * Return the next record to which this cursor points, and advance the cursor.
    *
    * @return Returns the record, default return array.
    *
    * @retval array   record
    * @retval string  record
    *
    * Example:
    * @code
    * $cursor = $cl -> find( array( 'a' => array( '$lte' => 50 ) ) ) ;
    * if( empty( $cursor ) ) {
    *    $err = $db -> getError() ;
    *    echo "Failed to find, error code: ".$err['errno'] ;
    *    return ;
    * }
    * while( $record = $cursor -> next() ) {
    *    var_dump( $record ) ;
    * }
    * @endcode
   */
   public function next(){}

   /**
    * Return the current element.
    *
    * @return Returns the record, default return array.
    *
    * @retval array   record
    * @retval string  record
    *
    * Example:
    * @code
    * $cursor = $cl -> find( array( 'a' => array( '$lte' => 50 ) ) ) ;
    * if( empty( $cursor ) ) {
    *    $err = $db -> getError() ;
    *    echo "Failed to find, error code: ".$err['errno'] ;
    *    return ;
    * }
    * if( $record = $cursor -> next() ) {
    *    var_dump( $record ) ;
    *    $record = $cursor -> current() ;
    *    var_dump( $record ) ;
    * }
    * @endcode
   */
   public function current(){}
}

File: driver/php/ext/class/class_cursor.c
/*******************************************************************************
   Copyright (C) 2012-2018 SequoiaDB Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*******************************************************************************/

#include "class_cursor.h"

extern INT32 connectionDesc ;
extern INT32 cursorDesc ;

PHP_METHOD( SequoiaCursor, __construct )
{
}

//e.g. Rename getNext
PHP_METHOD( SequoiaCursor, next )
{
   INT32 rc = SDB_OK ;
   zval *pThisObj = getThis() ;
   sdbCursorHandle cursor = SDB_INVALID_HANDLE ;
   bson record ;
   bson_init( &record ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    cursor,
                    sdbCursorHandle,
                    SDB_CURSOR_HANDLE_NAME,
                    cursorDesc ) ;
   bson_init( &record ) ;
   rc = sdbNext( cursor, &record ) ;
   if( rc )
   {
      goto error ;
   }
done:
   PHP_RETURN_AUTO_RECORD( FALSE,
                           pThisObj,
                           (rc == SDB_OK ? FALSE : TRUE),
                           record ) ;
   bson_destroy( &record ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}

PHP_METHOD( SequoiaCursor, current )
{
   INT32 rc = SDB_OK ;
   zval *pThisObj = getThis() ;
   sdbCursorHandle cursor = SDB_INVALID_HANDLE ;
   bson record ;
   bson_init( &record ) ;
   PHP_SET_ERRNO_OK( FALSE, pThisObj ) ;
   PHP_READ_HANDLE( pThisObj,
                    cursor,
                    sdbCursorHandle,
                    SDB_CURSOR_HANDLE_NAME,
                    cursorDesc ) ;
   bson_init( &record ) ;
   rc = sdbCurrent( cursor, &record ) ;
   if( rc )
   {
      goto error ;
   }
done:
   PHP_RETURN_AUTO_RECORD( FALSE,
                           pThisObj,
                           (rc == SDB_OK ? FALSE : TRUE),
                           record ) ;
   bson_destroy( &record ) ;
   return ;
error:
   PHP_SET_ERROR( FALSE, pThisObj, rc ) ;
   goto done ;
}